pkg/fleet-manager/application_controller.go

```diff
@@ -0,0 +1,384 @@
+	//  For example, if a cluster is added or removed from the fleet, should the application respond correspondingly?
+	//  Note that the relationship between fleet/cluster/attachedCluster and application is single-to-many, which can be complex to implement.
+	if err := c.Watch(
+		&source.Kind{Type: &fleetapi.Fleet{}},
+		handler.EnqueueRequestsFromMapFunc(a.objectToApplicationFunc),
+	); err != nil {
+		return fmt.Errorf("failed to add a Watch for GitRepository: %v", err)
+	}
+
+	// Set up watches for the updates to application's status.
+	// GitRepository/HelmRepository/Kustomization/HelmRelease and application have a many-to-single relationship.
+	if err := c.Watch(
+		&source.Kind{Type: &sourcev1.GitRepository{}},
+		handler.EnqueueRequestsFromMapFunc(a.objectToApplicationFunc),
+	); err != nil {
+		return fmt.Errorf("failed to add a Watch for GitRepository: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &sourcev1a2.HelmRepository{}},
+		handler.EnqueueRequestsFromMapFunc(a.objectToApplicationFunc),
+	); err != nil {
+		return fmt.Errorf("failed to add a Watch for HelmRepository: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kustomizev1.Kustomization{}},
+		handler.EnqueueRequestsFromMapFunc(a.objectToApplicationFunc),
+	); err != nil {
+		return fmt.Errorf("failed to add a Watch for Kustomization: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &helmv2b1.HelmRelease{}},
+		handler.EnqueueRequestsFromMapFunc(a.objectToApplicationFunc),
+	); err != nil {
+		return fmt.Errorf("failed to add a Watch for HelmRelease: %v", err)
+	}
+
+	return nil
+}
+
+func (a *ApplicationManager) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	app := &applicationapi.Application{}
+	if err := a.Get(ctx, req.NamespacedName, app); err != nil {
+		if apiserrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get application %s", req.NamespacedName)
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+	log = log.WithValues("application", klog.KObj(app))
+
+	patchHelper, err := patch.NewHelper(app, a.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to init patch helper for application %s", req.NamespacedName)
+	}
+
+	defer func() {
+		patchOpts := []patch.Option{}
+		if err := patchHelper.Patch(ctx, app, patchOpts...); err != nil {
+			reterr = utilerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch application %s", req.NamespacedName)})
+		}
+	}()
+
+	// Add finalizer if not exist to void the race condition.
+	if !controllerutil.ContainsFinalizer(app, ApplicationFinalizer) {
+		controllerutil.AddFinalizer(app, ApplicationFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	// Define object key for fleet based on the current policy's destination
+	fleetKey := client.ObjectKey{
+		Namespace: app.Namespace,
+		// there only one fleet
+		Name: app.Spec.SyncPolicy[0].Destination.Fleet,
```

Comment:
Assume only one syncpolicy is supported now if so add a TODO here

Reply:
support multi syncpolicy